### DISCUSSION TITLE
Channel/Sensor-Extension to measure even more circuits

### DISCUSSION THREAD

**Discussion by cocanut** [David00/rpi-power-monitor#97] [Ideas]
I would like to know if there is already some Ideas on the topic of extending the "monitor pcb" with 6 channels/sensors. As far as I understand I could theoretically setup 3 different "monitor pcbs" with 6 channels each. I would then connect each "monitor pcb" to one of the three phases in my household. Then I would need to figure out how all thee "monitor pcbs" can communicate via SPI with the Raspberry Pi without interfering with each other and finally I would need to adjust the "rpi-power-monitor" software to take into account even more channels/sensors. Is that correct?

**Comment by David00** [comment]
Hi @cocanut, I have been working on a new project, based on this one, that has 14 inputs on a single Pi.  I'm not quite ready to release it yet, but it is available via a presale beta.  Feel free to shoot me an email at github@dalbrecht.tech if you're interested!

**Reply by cocanut** [reply]
Hello David00,
thank you for your reply. I really would like to get into this topic. I am however still unsure whether it makes sense to jump into the rpi-power-monitor as for me it is still not completely clear how I could measure all three phases in my household reliable. As far as I understand, there is an error in the measurements on the two phases I do not measure the reference voltage because I suppose that the voltage is identical on each phase. The more the phases are loaded unevenly the more the error grows. Unfortunately I cannot determine what magnitude the error has, so I am not sure how to decide.

Referring to my original question in this topic: wouldn't it also be possible to put together 3 individual rpi-power-monitors each measuring one of the three phases? This way I would have 3 x 6 = 18 channels already. I would use 3 of them to directly measure the main electricity input, 3 to measure the 3-phase-circuit which is in charge of making warm water for showers. Then I still have 12 channels left to measure individual single phase consumers.

**Reply by David00** [reply]
> Referring to my original question in this topic: wouldn't it also be possible to put together 3 individual rpi-power-monitors each measuring one of the three phases?

Yep! And, you can send the measurements from each pi to a [remote InfluxDB server](https://david00.github.io/rpi-power-monitor/docs/v0.3.0-beta/advanced-usage.html#using-a-remote-influxdb-server), that way all the data goes to the same place and can be visualized with Grafana easily.  And, as of v0.3.0, the new `config.toml` allows you to [name the power monitor](https://david00.github.io/rpi-power-monitor/docs/v0.3.0-beta/configuration.html#name) - the name gets associated with all of the power monitor's data points, allowing you to distinguish which measurements came from which power monitor.

The one limitation with any 3-phase implementation is the way the AC voltage is connected to the board, which is via a 9V AC transformer. So even if you did have three separate power monitors, you'd need access to an outlet on each phase to connect three separate 9V AC transformers to.  This is likely another challenge for most people, because I don't think it's common to have outlets on all three phases right at the panel.

The ideal solution would be a variant of the power monitor PCB that can work directly with the voltage from each of the three phases, or, an expansion board that can be used along side the current version of the power monitor PCB.  I am definitely interested in exploring this idea more, but until then, using multiple monitors can accomplish the same thing.

As an aside, if you do deploy multiple monitors and setup a remote InfluxDB server to receive and store the data, and a separate Grafana server to visualize the data, you should be able to use a Pi Zero 2W for each monitor instead of a Pi 4. The Pi 4 is really only needed for its extra resources to run Grafana and InfluxDB right on the Pi. Let's hope the stock levels return to normal soon!!

**Reply by cocanut** [reply]
> The one limitation with any 3-phase implementation is the way the AC voltage is connected to the board, which is via a 9V AC transformer. So even if you did have three separate power monitors, you'd need access to an outlet on each phase to connect three separate 9V AC transformers to. This is likely another challenge for most people, because I don't think it's common to have outlets on all three phases right at the panel.

I agree that finding an outlet for each phase might be a difficult job. In my case I can install new outlets on my own, this might however not be an option for others. Thats also a good reason to try to estimate the error that might be done here. So how could I calculate the error? My idea would be something like:

True Voltage on phase 1: 225,35V
True Voltage on phase 2: 226,18V
Difference: (226,18-225,35) / 226,18 = 0,4%

To calculate the energy consumed on each phase I need to measure the current flowing through them. Lets say the measured current is exactly 10A on each phase. I would now calculate the energy consumed like this:

Energy consumed on phase 1: 225,35 * 10 = 2,2535 kW
Energy consumed on phase 2: 226,18 * 10 = 2,2618 kW
Absolute error for phase 2: 2,2618 - 2,2535 = 0,0083 kW
Relative error for phase 2: 0,0083/2,2618 = approx. 0,4%

Unsurprinsingly the error of 0,4% for the energy consumed is the same for the voltage measurement. I think 0,4% would be really no big deal. Is there any experience how much the voltage can diverge in reality?

I know in germany the potential drop from the mains panel to the consumer is allowed to be max. 5% but that is more about the length and the cross section of the cables from the panel to the consumer. But I think there is also a change in the potential due to unsymmetric loading of the three phases? Unfortunately I am no expert here so please tell my if I am wrong.

**Reply by David00** [reply]
I actually do have a small sample set of data from a 3-phase system in Australia.  The individual I was working with took manual voltage measurements frequently (about 12 times) over a single day, and found that the voltage across all three phases varied by as much as 7V.  So, that can be quite a big error in the resulting power calculation.

Furthermore, the simple current * voltage power calculation gives you the _apparent_ power in volt-amperes (VA). The "real power" calculation (which gives you Watts) considers the actual phase angle between the voltage samples and the current samples.  Without access to voltage measurements from each of the three phases, some assumptions have to be made about the phase angle relationship on the two phases where we don't get any voltage samples.  My 3-phase variant accounts for the phase angle problem fairly well but it can't do anything about the voltage variances.  We'd have to get voltage samples from all three phases to truly be accurate.

**Reply by cocanut** [reply]
I see ... I am missing a whole lot on electronics therory to estimate this error, thank you for the clarification! The 7V you have measured are already way more than I expected.

**Reply by vulture2600** [reply]
Is it possible to stack multiple hats to give more inputs? Changing the SPI address might be challenging but if that was the only thing preventing someone from being able to do this, it might be worth looking into.

**Reply by David00** [reply]
@vulture2600 Unfortunately it's not possible to stack multiple boards without a PCB hardware revision.  I have some concerns about the sampling rate too, but it's looking like the new v6 kernels have much better throughput over SPI, so that might be promising.
